python/tk_desktop/rpc.py

```diff
@@ -268,3 +393,540 @@ def close(self):
+
+            # execute the function on the main thread, as it may to GUI work.
+            if is_blocking:
+                response = self._engine.execute_in_main_thread(func, *args, **kwargs)
+            else:
+                self._engine.async_execute_in_main_thread(func, *args, **kwargs)
+                response = None
+
+            logger.debug("http '%s' result: '%s" % (func_name, response))
+        except Exception as e:
+            response = e
+            # if any of the above fails send the exception back to the client
+            logger.exception("got exception during '%s" % func_name)
+        finally:
+            self._send_response(response)
+
+    def log_request(self, code=None, size=None):
+        """
+        Prevent the server from outputting to stdout every single time
+        a query is processed by the server.
+        """
+
+    def _read_request(self):
+        """
+        Read an incoming request and return the data.
+
+        :returns: The data that was sent.
+        """
+        if six.PY3:
+            content_len = int(self.headers.get("content-length", 0))
+        else:
+            content_len = int(self.headers.getheader("content-length", 0))
+
+        body = self.rfile.read(content_len)
+        try:
+            return pickle.loads(body)
+        except Exception:
+            raise RuntimeError("Response could not be unserialized: %s" % body)
+
+    def _send_response(self, body):
+        """
+        Send a response back to the client.
+
+        :param body: Data to send back to the client.
+        """
+        raw_data = pickle.dumps(body)
+        self.send_response(200, self.responses[200])
+        self.send_header("Content-Type", "text/text")
+        self.send_header("Content-Length", len(raw_data))
+        self.send_header("Connection", "close")
+        self.end_headers()
+        self.wfile.write(six.ensure_binary(raw_data))
+
+
+######################
+# Client side classes
+class Connection(object):
+    """
+    Connection to the other process's server.
+    """
+
+    def __init__(self, address, authkey):
+        """
+        :param str address: Address of the server.
+        :param str authkey: Credentials to the server.
+        """
+        self._address = address
+        self._authkey = authkey
+
+    def send(self, is_blocking, func_name, args, kwargs):
+        """
+        Call a function on the server synchronously or not.
+
+        :param bool is_blocking: If ``True``, the call will block until the server
+            completes the task and the value/exception will be returned/raised by this
+            method. If ``False``, the call will be queued for delivery and the method
+            returns None.
+        """
+        payload = [is_blocking, func_name, args, kwargs]
+        payload = pickle.dumps([self._authkey, payload])
+        # if is_blocking:
+        request = urlopen(self._address, data=six.ensure_binary(payload))
```

Comment:
Is this line adding it to the queue?

I'm clear on the non blocking behaviour here, it seems we create a request and then close it straight away?

Reply:
I've cleared up the confusion.